turbo/jsonrpc/eth_call.go

```diff
@@ -334,100 +337,150 @@ func (api *APIImpl) EstimateGas(ctx context.Context, argsOrNil *ethapi2.CallArgs
+	}
+
+	// generate the trie for proofs, this works by loading the merkle paths to the touched keys
+	proofTrie, proofRootHash, err := sdCtx.GenerateTouchedKeyTrie(ctx, header.Root[:], "eth_getProof")
+	if err != nil {
+		return nil, err
+	}
+
+	// verify hash
+	if !bytes.Equal(proofRootHash, header.Root[:]) {
+		return nil, fmt.Errorf("proof root hash mismatch actual(%x)!=expected(%x)", proofRootHash, header.Root[:])
+	}
+
+	a := new(big.Int).SetInt64(0)
+	// set initial response fields
+	proof := &accounts.AccProofResult{
+		Address:      address,
+		Balance:      (*hexutil.Big)(a),
+		Nonce:        hexutil.Uint64(0),
+		CodeHash:     libcommon.Hash{},
+		StorageHash:  libcommon.Hash{},
+		StorageProof: make([]accounts.StorProofResult, len(storageKeys)),
+	}
+
+	// get account proof
+	accountProof, _, err := proofTrie.Prove(crypto.Keccak256(address.Bytes()), 0, false)
+	if err != nil {
+		return nil, err
+	}
+	proof.AccountProof = *(*[]hexutility.Bytes)(unsafe.Pointer(&accountProof))
+
+	// get account data from the trie
+	acc, _ := proofTrie.GetAccount(crypto.Keccak256(address.Bytes()))
+	if acc == nil {
+		for i, k := range storageKeys {
+			proof.StorageProof[i] = accounts.StorProofResult{
+				Key:   k,
+				Value: (*hexutil.Big)(a),
+				Proof: nil,
 			}
-			tx = batch
-		} else {
-			loader = trie.NewFlatDBTrieLoader("eth_getProof", rl, nil, nil, false)
 		}
+		return proof, nil
+	} else {
+		proof.Balance = (*hexutil.Big)(acc.Balance.ToBig())
+		proof.Nonce = hexutil.Uint64(acc.Nonce)
+		proof.CodeHash = acc.CodeHash
+		proof.StorageHash = acc.Root
+	}
 
-		reader, err := rpchelper.CreateStateReader(ctx, tx, blockNrOrHash, 0, api.filters, api.stateCache, "")
-		if err != nil {
-			return nil, err
-		}
-		a, err := reader.ReadAccountData(address)
-		if err != nil {
-			return nil, err
-		}
-		if a == nil {
-			a = &accounts.Account{}
-		}
-		pr, err := trie.NewProofRetainer(address, a, storageKeys, rl)
+	// get storage key proofs
+	for i, keyHash := range storageKeys {
+		// prepare key path (keccak(address) | keccak(key))
+		var fullKey []byte
+		fullKey = append(fullKey, crypto.Keccak256(address.Bytes())...)
+		fullKey = append(fullKey, crypto.Keccak256(keyHash.Bytes())...)
+
+		// get proof for the given key
+		storageProof, value, err := proofTrie.Prove(fullKey, len(proof.AccountProof), true)
 		if err != nil {
-			return nil, err
+			return nil, errors.New("cannot verify store proof")
 		}
 
-		loader.SetProofRetainer(pr)
-		root, err := loader.CalcTrieRoot(tx, nil)
-		if err != nil {
-			return nil, err
+		// Decode the hexadecimal string into the big.Int
+		// The base is 16 for hexadecimal
+		n := new(big.Int)
+		n.SetString(hex.EncodeToString(value), 16)
+
+		// set key proof
+		proof.StorageProof[i].Key = keyHash
+		proof.StorageProof[i].Value = (*hexutil.Big)(unsafe.Pointer(n))
+
+		// 0x80 represents RLP encoding of an empty proof slice
+		if proof.StorageHash.Cmp(libcommon.BytesToHash(commitment.EmptyRootHash)) != 0 {
```

Comment:
Can move this above `proofTrie.Prove` to avoid calling Prove in this case.

Reply:
Oh but I guess you are relying on `Prove` to get the value, so it might not be possible to skip that call.